### ISSUE TITLE
[Core]Read working_dir zip from private s3 

### ISSUE THREAD

**Issue by loleek** [ray-project/ray#34708]
### Description

Currently, boto3 cannot inject endpoint_url from ~/.aws/config or env_vars.
So I cannot submit a job which located at my private s3-like storage using s3-prefix working_dir.
I see the packing.py create boto3.client very simple:
```
 if protocol == Protocol.S3:
       try:
           import boto3
           from smart_open import open as open_file
       except ImportError:
           raise ImportError(
               "You must `pip install smart_open` and "
               "`pip install boto3` to fetch URIs in s3 "
               "bucket. " + install_warning
           )
       tp = {"client": boto3.client("s3")}
```
Can ray expose kwags here so I can inject endpoint_url, access_key_id, access_secret or other configurations?

**Comment by rkooo567** [comment]
Hmm this seems to be a pretty reasonable request. cc @architkulkarni for thoughts?

**Comment by architkulkarni** [comment]
I agree, it seems reasonable. It may take some care to get the API right.

**Comment by rkooo567** [comment]
This is as similar issue as https://github.com/ray-project/ray/issues/31122.

**Comment by ketangangal** [comment]
Hello @rkooo567 & @architkulkarni 

can i work on this issue ?

**Comment by ketangangal** [comment]
Requirements : 
1. if From Kwargs **endpoint_url, access_key_id, access_secret or other configurations?**
2. elif From ENV **endpoint_url, access_key_id, access_secret or other configurations?**
3. else. default boto3

**Comment by rkooo567** [comment]
that sounds great! @rynewang can shepherd the contribution!

**Comment by rynewang** [comment]
The same mechanism is used by `java_jars`, `py_modules` and `working_dir`. We can find a common way to model these.

We can either type the `working_dir` value to (pseudocode) `str | Dict["uri" | **kwargs, str]`, accepting

- `runtime_env={"working_dir":{"uri":"s3://my_private_bucket/dir", "aws_access_key_id":"my_key", "aws_secret_access_key":"my_secret"}}`
- `runtime_env={"working_dir":"s3://public_bucket/dir"}`

Or we can add a special field in runtime_env `_runtime_env_context` like

- `runtime_env={"working_dir":"s3://my_private_bucket/dir", "_runtime_env_context":{"aws_access_key_id":"my_key", "aws_secret_access_key":"my_secret"}}`

and let different runtime env plugins to read them.

I personally prefer the former one, but I'd like to hear from you. cc @jjyao @rkooo567

**Comment by architkulkarni** [comment]
I believe `boto3` now supports specifying the endpoint URL via environment variable (which was the initial reason for creating this issue).  See https://github.com/boto/boto3/issues/2099#issuecomment-1626117882

If we can verify that this workflow can succeed by setting the relevant variables in the `runtime_env` `"env_vars"` field, then perhaps we don't even need to add a new API here.  Instead we can just add a quick section for this workflow in the docs.

**Comment by stwerner97** [comment]
@architkulkarni I've tried to do this for the ``ServeSubmissionClient`` where I set the ``AWS_ACCESS_KEY_ID``, ``AWS_SECRET_ACCESS_KEY``, and  ``AWS_ENDPOINT_URL`` in the ``env_vars`` field of ``runtime_env`` but get an ``botocore.exceptions.NoCredentialsError: Unable to locate credentials`` exception when Ray attempts to download the ``working_dir`` contents (remote URI ``s3://``) . I've checked that the environment variables are passed correctly to the application ( I can access my S3 bucket with custom boto calls ).

**Comment by Blackoutta** [comment]
> [@architkulkarni](https://github.com/architkulkarni) I've tried to do this for the `ServeSubmissionClient` where I set the `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_ENDPOINT_URL` in the `env_vars` field of `runtime_env` but get an `botocore.exceptions.NoCredentialsError: Unable to locate credentials` exception when Ray attempts to download the `working_dir` contents (remote URI `s3://`) . I've checked that the environment variables are passed correctly to the application ( I can access my S3 bucket with custom boto calls ).

any update on this? I'm stuck here too.
